### DISCUSSION TITLE
Our simulation consumes too much RAM.

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3399] [Q&A]
I am simulating Tor networks that are generated by `tornettools`. I am moving from 2% networks to 10% ones, as recommended at the end of the "Once is Never Enough" paper ([link](https://www.robgjansen.com/publications/neverenough-sec2021.pdf)). I hope to reach a simulated time of 3 hours.

But even when I use the vanilla Tor network generated by tornettools via the command
```
tornettools generate     relayinfo_staging_2023-04-01--2023-04-30.json     userinfo_staging_2023-04-01--2023-04-30.json     networkinfo_staging.gml     tmodel-ccs2018.github.io     --network_scale 0.1     --prefix tornets/original/0.1_10800s/2
```
without any further modification (except the stop time), the simulation consumed all of the 306G of memory + 8G of swap available to me, and exited with code -9 at the 1.5 hour mark.

Comparing my simulation with Table 4 in the "Once is Never Enough" paper, I see that my simulation consumed about 200G of RAM after 1 hour of simulated time, while the paper shows a 355G of RAM consumption for a 10% network and 1 hour stop time. So there is a difference between the two. I'd like to ask if a 200G of RAM consumption in my case is normal?

What concerns me more is that if I want to simulate for 3 hours, given the linear growth of RAM usage shown in the plot, it seems that I'll need twice as much RAM, which is about 600G. My lab and I could reasonably afford some more RAM. But the bigger problem is that my main experiment requires adding a new process to about 20% of the clients in the network, which will be gathering data and talking to a new bandwidth authority. At the previous 2% scale and 1 hour stop time, the original network's RAM usage grew linearly and ended up at around 38G after the 1-hour simulation is done. The main experiment ended at around 50G, but had brief but sharp spikes to 65G.

So my second question is how to estimate the memory usage for a given setup? We want to avoid buying more RAM but find out later that they are not enough and we can't afford more RAM either. Can I use `tornettools` to generate a 10% network, set it up for our main experiement, and set the stop time to be 3 hours. When it hits the memory limit and exits, can I use the simulated time and exit time to estimate the total RAM needed for 3 hours? E.g. If the simulation consumes 300G of RAM and exits at 1 hour of simulated time, can I typically expect that it will consume 300G * 3 = 900G of RAM if I want to simulate it for 3 hours? And how to deal with RAM spikes? The machine has multiple users. Could the spikes be coming from other users' workload (not related to Shadow)?

My third question is how can I optimize the simulation for RAM usage? I'm looking at the "Performance Tuning" section in the [docs](https://shadow.github.io/docs/guide/sidechannels.html). We are not adding "mitigations=off" to the kernel parameters to disbale side-channel mitigations at the moment. The "Configuration options" section seems the most accessible to me right now. But I only find `heartbeat_interval` and `host_heartbeat_interval` to be somewhat safe to increase. Though it's unclear to me if there are any negative effects, e.g. will the plot be more coarse? The `runahead` option may be relevant. Since I'm simulating a Tor network, will it be reasonable if I increase `runahead` to be bigger than the default 1ms, because the latency between Tor nodes are generally bigger than 1ms? Or will that be a mistake? And in general, it's unclear to me whether these 3 options, and other options can reduce the memory usage, or that they are more for reducing the run time.

I have attached the plot for the 10% 3-hour simulation of the original network.

[tornet.plot.pages.pdf](https://github.com/user-attachments/files/16842043/tornet.plot.pages.pdf)

**Comment by stevenengler** [comment]
I don't run Tor simulations so I can't really offer any advice, but I'll mention that the "Once is never enough" paper used Shadow 1.x, which was a very different simulator compared to the later ones (2.x and onwards). The old version used a single-process architecture, and it's possible that RAM usage scales differently between the two designs.

In general Shadow doesn't have any way to tune the memory usage of a simulation. Maybe someone has come up with some things to tweak, but this isn't something that I'm aware of or that we've documented somewhere.

As far as I know, we don't know exactly where the memory is being used or why it increases linearly. It might be normal, or maybe there are some queues in shadow that get very long, or maybe there's a memory leak somewhere. It would be useful to investigate this further, but I don't know of anyone who has looked into it deeply. We've tried looking for memory leaks a few times but didn't find anything.

In the appendix of the "Co-opting Linux Processes for High-Performance Network Simulation" paper, there's a graph and a short mention of memory usage that says that the new (at the time) Shadow simulator (2.x) is slightly more memory efficient than the old Shadow simulator (1.x), but I don't know any of the details, and the current simulator (3.x) might be different too.

**Reply by higher-order** [reply]
Thanks for pointing out the differences among Shadow versions. And I thought the growth of memory usage is normal, maybe due to Shadow needing to keep track of states. But it seems to be not the case now then.

**Reply by stevenengler** [reply]
I wouldn't say that the linear growth is not normal. More that it seems suspicious and is worth some more investigation to see if there's anything more that can be done to improve it.

**Comment by sporksmith** [comment]
Unfortunately we haven't really been able to pin down the linear growth in RAM usage. As near as I could tell when I looked into this, the growth was primarily in the `tor` processes. `tor` tries to limit its total RAM usage, and I *think* we looked before to ensure each process isn't trying to consume the whole machine's RAM, but that's probably where I'd resume investigation if I were to pick this up again.

Setting `TestingEnableCellStatsEvent 0` in the `tor` config files helps a fair bit. `tornettools` already does this by default though, as of [aa6360930dbfaad16929eed6fd188b5ffa67da47](https://github.com/shadow/tornettools/commit/aa6360930dbfaad16929eed6fd188b5ffa67da47)

**Reply by higher-order** [reply]
I thought the growth of memory usage is normal, maybe because Shadow needs to keep track of all previous states. I guess that's not the case then, given your and Steven's responses.

**Reply by robgjansen** [reply]
So we've known about the linear increase in memory usage when running Tor experiments for some time. Whether or not the increase in memory usage should be considered "normal" depends how you define "normal". Relays running in the live Tor network can easily consume 1 GiB of memory or more. With a 10% network, you have about 600 relays, and if each eventually consumes 1 GiB, you'll get up to 600 GiB fairly quickly. I think this is what is happening in the Shadow simulations.

Some of the questions we have are:
1. How should we expect a relay's memory usage to grow over time as it starts getting more and more usage? That is, if we ran a Tor relay outside of shadow and started sending lots of connections through it, how does the memory grow/change over time? Does the memory usage eventually "top out" or reach a maximum that it does not really exceed, and if so what is the maximum?
2. Where exactly is the memory being consumed in the Tor process? Are various things being cached? Is it storing stats or other history data? Is it just that more circuits in parallel means more queues and more memory?
3. Is there a problem/peculiarity that is specific to Shadow that is causing the memory usage? For example, are some TCP connections not getting closed properly, which might keep those sockets active in Tor and cause their buffers to stick around longer than expected?

We might be able to get a handle on some of this by running a smaller scale experiment for several hours of simulation time. It would be interesting if during such an experiment we observed a linear increase and then a plateau of memory usage.

**Reply by higher-order** [reply]
How small the scale of the network can be for the experiment you mentioned? Can a 0.1% network with just a few relays be helpful?

**Reply by robgjansen** [reply]
I think I would aim for a 1% network if possible. We want to make sure that we have enough memory usage that we'll be able to notice if a change in ram usage is statistically significant, and if we use a network that is too small it may be difficult to isolate the memory usage to the shadow simulation. But this is just a guess, feel free to play around with other sizes too! :)
